lib/ruby_smb/gss/provider/ntlm.rb

```diff
@@ -0,0 +1,337 @@
+            :IDENTIFY                 => 1 << 20,
+            :NON_NT_SESSION           => 1 << 22,
+            :TARGET_INFO              => 1 << 23,
+            :VERSION_INFO             => 1 << 25,
+            :KEY128                   => 1 << 29,
+            :KEY_EXCHANGE             => 1 << 30,
+            :KEY56                    => 1 << 31
+          }.freeze
+
+          def reset!
+            super
+            @server_challenge = nil
+          end
+
+          def process(request_buffer=nil)
+            if request_buffer.nil?
+              # this is only NTLMSSP (as opposed to SPNEGO + NTLMSSP)
+              buffer = OpenSSL::ASN1::ASN1Data.new([
+                Gss::OID_SPNEGO,
+                OpenSSL::ASN1::ASN1Data.new([
+                  OpenSSL::ASN1::Sequence.new([
+                    OpenSSL::ASN1::ASN1Data.new([
+                      OpenSSL::ASN1::Sequence.new([
+                        Gss::OID_NTLMSSP
+                      ])
+                    ], 0, :CONTEXT_SPECIFIC),
+                    OpenSSL::ASN1::ASN1Data.new([
+                      OpenSSL::ASN1::ASN1Data.new([
+                        OpenSSL::ASN1::ASN1Data.new([
+                          OpenSSL::ASN1::GeneralString.new('not_defined_in_RFC4178@please_ignore')
+                        ], 0, :CONTEXT_SPECIFIC)
+                      ], 16, :UNIVERSAL)
+                    ], 3, :CONTEXT_SPECIFIC)
+                  ])
+                ], 0, :CONTEXT_SPECIFIC)
+              ], 0, :APPLICATION).to_der
+              return Result.new(buffer, WindowsError::NTStatus::STATUS_SUCCESS)
+            end
+
+            begin
+              gss_api = OpenSSL::ASN1.decode(request_buffer)
+            rescue OpenSSL::ASN1::ASN1Error
+              return
+            end
+
+            if gss_api&.tag == 0 && gss_api&.tag_class == :APPLICATION
+              result = process_gss_type1(gss_api)
+            elsif gss_api&.tag == 1 && gss_api&.tag_class == :CONTEXT_SPECIFIC
+              result = process_gss_type3(gss_api)
+            end
+
+            result
+          end
+
+          #
+          # Process the NTLM type 1 message and build a type 2 response message.
+          #
+          # @param [Net::NTLM::Message::Type1] type1_msg the NTLM type 1 message received by the client that should be
+          #   processed
+          # @return [Net::NTLM::Message::Type2] the NTLM type 2 response message with which to reply to the client
+          def process_ntlm_type1(type1_msg)
+            type2_msg = Net::NTLM::Message::Type2.new.tap do |msg|
+              msg.target_name = 'LOCALHOST'.encode('UTF-16LE').b
+              msg.flag = 0
+              %i{ KEY56 KEY128 KEY_EXCHANGE UNICODE TARGET_INFO VERSION_INFO }.each do |flag|
+                msg.flag |= NEGOTIATE_FLAGS.fetch(flag)
+              end
+
+              @server_challenge = @provider.generate_server_challenge
+              msg.challenge = @server_challenge.unpack1('Q')
```

Comment:
Comments on pack always helps with the eyebleed, something like:
`Q => 64-bit unsigned, native endian (uint64_t)`